website/sales/models/shift.py

```diff
@@ -0,0 +1,148 @@
+
+    locked = models.BooleanField(
+        blank=False,
+        null=False,
+        default=False,
+        help_text=_(
+            "Prevent orders being changed or created for this shift. This will also clean up all unpaid orders in this shift."
+        ),
+    )
+
+    def clean(self):
+        super().clean()
+        errors = {}
+
+        if self.end and self.start and self.end <= self.start:
+            errors.update({"end": _("End cannot be before start.")})
+
+        if errors:
+            raise ValidationError(errors)
+
+    def save(
+        self, force_insert=False, force_update=False, using=None, update_fields=None
+    ):
+        if self.locked:
+            self.orders.filter(
+                (Q(payment__isnull=True) & Q(total_amount__gt=0))
+                | Q(order_items__isnull=True)
+            ).delete()
+
+        return super(Shift, self).save(force_insert, force_update, using, update_fields)
+
+    active = RangeCheckProperty("start", "end", timezone.now)
+
+    @queryable_property(annotation_based=True)
+    @classmethod
+    def total_revenue(cls):
+        return RawSQL(
+            """(SELECT CAST(COALESCE(SUM("total__"), 0) AS NUMERIC) AS "shift_revenue__"
+                FROM (
+                    SELECT "sales_order"."id", "sales_order"."shift_id", "sales_order"."discount", "sales_order"."payment_id", CAST(SUM("total") AS NUMERIC) AS "subtotal__", CAST((SUM("total") - COALESCE("discount", 0)) AS NUMERIC) AS "total__", SUM("amount") AS "num_items__"
+                    FROM "sales_order" LEFT JOIN "sales_orderitem" ON "sales_orderitem"."order_id" = "sales_order"."id"
+                    GROUP BY "sales_order"."id", "sales_order"."shift_id", "sales_order"."discount"
+                    )
+                WHERE "shift_id"="sales_shift"."id"
+                )""",
+            [],
+        )
+
+    @queryable_property(annotation_based=True)
+    @classmethod
+    def total_revenue_paid(cls):
+        return RawSQL(
+            """(SELECT CAST(COALESCE(SUM("total__"), 0) AS NUMERIC) AS "shift_revenue__"
+                FROM (
+                    SELECT "sales_order"."id", "sales_order"."shift_id", "sales_order"."discount", "sales_order"."payment_id", CAST(SUM("total") AS NUMERIC) AS "subtotal__", CAST((SUM("total") - COALESCE("discount", 0)) AS NUMERIC) AS "total__", SUM("amount") AS "num_items__"
+                    FROM "sales_order" LEFT JOIN "sales_orderitem" ON "sales_orderitem"."order_id" = "sales_order"."id"
+                    GROUP BY "sales_order"."id", "sales_order"."shift_id", "sales_order"."discount"
+                    )
+                WHERE "shift_id"="sales_shift"."id"
+                AND ("payment_id" IS NOT NULL OR ("payment_id" IS NULL AND "total__"=0))
+                )""",
+            [],
+        )
+
+    @queryable_property(annotation_based=True)
+    @classmethod
+    def num_orders(cls):
+        return RawSQL(
+            """(SELECT COUNT(*) FROM "sales_order"
```

Comment:
This really looks like something you should not need raw SQL for.

Reply:
Agreed